src/mpid/ch4/netmod/ofi/ofi_win.h

```diff
@@ -56,13 +56,65 @@ static inline int MPIDI_OFI_win_progress_fence(MPIR_Win * win)
 
     r = MPIDI_OFI_WIN(win).syncQ;
 
-    while (r) {
-        MPIDI_OFI_win_request_t *next = r->next;
-        MPIDI_OFI_rma_done_event(NULL, (MPIR_Request *) r);
-        r = next;
+    /* Should not free the RMA request when do_free = 0; manual RMA flush could happen in
+     * the middle of issuing one MPI-level RMA op and it could lead to premature freeing of
+     * the request. */
+    if (do_free) {
+        while (r) {
+            MPIDI_OFI_win_request_t *next = r->next;
+            MPIDI_OFI_rma_done_event(NULL, (MPIR_Request *) r);
+            r = next;
+        }
+
+        MPIDI_OFI_WIN(win).syncQ = NULL;
+        MPIDI_OFI_WIN(win).counter = 1;
+    }
```

Comment:
I guess it's not incorrect, but it could skew the results toward calling the progress engine more often than it could be. I guess this was being conservative.